File: lib/models/earthquake.dart
import 'package:xml/xml.dart';

class Earthquake {
  final String id;
  final String title;
  final DateTime date;
  final double mag;
  final double depth;
  final double latitude;
  final double longitude;
  double distance; // Metre veya km cinsinden aradaki mesafe

  Earthquake({
    required this.id,
    required this.title,
    required this.date,
    required this.mag,
    required this.depth,
    required this.latitude,
    required this.longitude,
    this.distance = 0.0,
  });

  factory Earthquake.fromJsonAfad(Map<String, dynamic> json) {
    double lat = 0.0;
    double lng = 0.0;
    
    if (json['geojson'] != null && json['geojson']['coordinates'] != null) {
      var coords = json['geojson']['coordinates'];
      if (coords is List && coords.length >= 2) {
        lng = _parseDouble(coords[0]);
        lat = _parseDouble(coords[1]);
      }
    } else {
      lat = _parseDouble(json['lat'] ?? json['latitude']);
      lng = _parseDouble(json['lng'] ?? json['longitude']);
    }

    return Earthquake(
      id: json['earthquake_id'] ?? json['id'] ?? '',
      title: json['title'] ?? json['location'] ?? 'Bilinmeyen Konum',
      date: _parseAfadDate(json['date_time'] ?? json['date']),
      mag: _parseDouble(json['mag'] ?? json['magnitude']),
      depth: _parseDouble(json['depth']),
      latitude: lat,
      longitude: lng,
    );
  }

  static DateTime _parseAfadDate(dynamic dateStr) {
    if (dateStr == null) return DateTime.now();
    try {
      // Format: "21.03.2024 13:00:24" veya "2024-03-21T13:00:24"
      if (dateStr.toString().contains('.')) {
        // DD.MM.YYYY HH:MM:SS formatı için basit parse
        final parts = dateStr.toString().split(' ');
        final dateParts = parts[0].split('.');
        final y = int.parse(dateParts[2]);
        final m = int.parse(dateParts[1]);
        final d = int.parse(dateParts[0]);
        final t = parts[1].split(':');
        return DateTime(y, m, d, int.parse(t[0]), int.parse(t[1]), int.parse(t[2]));
      }
      return DateTime.tryParse(dateStr.toString())?.toLocal() ?? DateTime.now();
    } catch (_) {
      return DateTime.now();
    }
  }

  static double _parseDouble(dynamic value) {
    if (value is int) {
      return value.toDouble();
    } else if (value is double) {
      return value;
    } else if (value is String) {
      return double.tryParse(value) ?? 0.0;
    }
    return 0.0;
  }

  // Kandilli API'den gelen veriyi parse eder (ESKİ JSON API - Orhan Aydoğdu)
  factory Earthquake.fromJsonKandilli(Map<String, dynamic> json) {
    List<dynamic> coordinates = json['geojson']['coordinates'] ?? [0.0, 0.0];
    // Kandilli API returns [longitude, latitude]
    double lng = _parseDouble(coordinates[0]);
    double lat = _parseDouble(coordinates[1]);

    return Earthquake(
      id: json['earthquake_id'] ?? json['_id'] ?? '',
      title: json['title'] ?? 'Bilinmeyen Konum',
      date: DateTime.tryParse(json['date'] ?? '')?.toLocal() ?? DateTime.now(),
      mag: _parseDouble(json['mag']),
      depth: _parseDouble(json['depth']),
      latitude: lat,
      longitude: lng,
    );
  }

  // Kandilli'nin orjinal XML (RSS) servisinden gelen veriyi parse eder
  factory Earthquake.fromKandilliXml(XmlElement item) {
    String titleText = item.findElements('title').isNotEmpty ? item.findElements('title').first.innerText : 'Bilinmeyen Konum';
    String descText = item.findElements('description').isNotEmpty ? item.findElements('description').first.innerText : '';
    
    // "1.5 (ML) MALAZGIRT" -> Büyüklüğü ayır
    double mag = 0.0;
    final magMatch = RegExp(r'^([\d\.]+)').firstMatch(titleText);
    if (magMatch != null) {
      mag = double.tryParse(magMatch.group(1) ?? '0') ?? 0.0;
    }

    String location = titleText.replaceAll(RegExp(r'^[\d\.]+\s*(?:\([a-zA-Z]+\)|\-)?\s*'), '').trim();

    // Tarihi açıklama formatından çıkar: "2024.12.11 09:22:06"
    DateTime date = DateTime.now();
    try {
      final dateMatch = RegExp(r'(\d{4})[./-](\d{2})[./-](\d{2})\s+(\d{2}):(\d{2}):(\d{2})').firstMatch(descText);
      if (dateMatch != null) {
        date = DateTime(
          int.parse(dateMatch.group(1)!),
          int.parse(dateMatch.group(2)!),
          int.parse(dateMatch.group(3)!),
          int.parse(dateMatch.group(4)!),
          int.parse(dateMatch.group(5)!),
          int.parse(dateMatch.group(6)!),
        );
      }
    } catch (_) {}

    // Derinlik ve Koordinatları çıkar
    double depth = 0.0;
    double lat = 0.0;
    double lng = 0.0;

    final depthMatch1 = RegExp(r'Derinlik[^\d]*([\d\.]+)').firstMatch(descText);
    if (depthMatch1 != null) {
      depth = double.tryParse(depthMatch1.group(1) ?? '0') ?? 0.0;
    }

    if (item.findElements('geo:lat').isNotEmpty) lat = double.tryParse(item.findElements('geo:lat').first.innerText) ?? 0.0;
    if (item.findElements('geo:long').isNotEmpty) lng = double.tryParse(item.findElements('geo:long').first.innerText) ?? 0.0;

    // Farklı formattaysa Description'dan arta kalan floatları topla: [mag, lat, lng, depth]
    if (lat == 0.0 || lng == 0.0 || depth == 0.0) {
      final floatMatches = RegExp(r'\b(\d+\.\d+)\b').allMatches(descText).toList();
      if (lat == 0.0 && floatMatches.length >= 3) {
        lat = double.tryParse(floatMatches[1].group(1) ?? '0') ?? lat;
        lng = double.tryParse(floatMatches[2].group(1) ?? '0') ?? lng;
      }
      if (depth == 0.0 && floatMatches.isNotEmpty) {
        depth = double.tryParse(floatMatches.length >= 4 ? floatMatches[3].group(1) ?? '0' : floatMatches.last.group(1)!) ?? 0.0;
      }
    }

    String link = item.findElements('link').isNotEmpty ? item.findElements('link').first.innerText : '';
    String id = link.isNotEmpty ? link.split('/').last.replaceAll('.asp', '') : date.millisecondsSinceEpoch.toString();

    return Earthquake(
      id: id,
      title: location.isEmpty ? 'Bilinmeyen Konum' : location,
      date: date,
      mag: mag,
      depth: depth,
      latitude: lat,
      longitude: lng,
    );
  }

  // USGS API'den gelen GeoJSON verisini parse eder
  factory Earthquake.fromJsonUSGS(Map<String, dynamic> json) {
    final properties = json['properties'];
    final geometry = json['geometry'];
    
    // USGS returns coordinates as [longitude, latitude, depth]
    List<dynamic> coordinates = geometry['coordinates'] ?? [0.0, 0.0, 0.0];
    
    // Time is in milliseconds since epoch
    final timeMs = properties['time'] as int?;
    DateTime date = timeMs != null 
        ? DateTime.fromMillisecondsSinceEpoch(timeMs).toLocal() 
        : DateTime.now();
        
    String title = properties['place'] ?? 'Bilinmeyen Konum';
    // Remove "X km Y of " prefix from USGS title to make it cleaner
    if (title.contains(' of ')) {
      title = title.split(' of ').last;
    }

    return Earthquake(
      id: json['id'] ?? '',
      title: title.trim(),
      date: date,
      mag: _parseDouble(properties['mag']),
      depth: _parseDouble(coordinates.length > 2 ? coordinates[2] : 0.0), // USGS provides depth as 3rd coordinate
      latitude: _parseDouble(coordinates[1]),
      longitude: _parseDouble(coordinates[0]),
    );
  }
}

File: depremanlik-kmp/shared/src/commonMain/kotlin/com/seismoalert/models/Earthquake.kt
package com.seismoalert.models

import kotlinx.datetime.Instant
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.Serializable
import kotlin.math.*

@Serializable
data class Earthquake(
    val id: String,
    val title: String,
    val dateEpochMs: Long,
    val mag: Double,
    val depth: Double,
    val latitude: Double,
    val longitude: Double,
    var distance: Double = 0.0 // Metre cinsinden mesafe
) {
    val formattedDate: String
        get() {
            return try {
                val instant = Instant.fromEpochMilliseconds(dateEpochMs)
                val dt = instant.toLocalDateTime(TimeZone.currentSystemDefault())
                "${dt.dayOfMonth.toString().padStart(2, '0')}.${dt.monthNumber.toString().padStart(2, '0')}.${dt.year} ${dt.hour.toString().padStart(2, '0')}:${dt.minute.toString().padStart(2, '0')}"
            } catch (_: Exception) {
                ""
            }
        }

    val formattedDateFull: String
        get() {
            return try {
                val instant = Instant.fromEpochMilliseconds(dateEpochMs)
                val dt = instant.toLocalDateTime(TimeZone.currentSystemDefault())
                "${dt.dayOfMonth.toString().padStart(2, '0')}.${dt.monthNumber.toString().padStart(2, '0')}.${dt.year} - ${dt.hour.toString().padStart(2, '0')}:${dt.minute.toString().padStart(2, '0')}:${dt.second.toString().padStart(2, '0')}"
            } catch (_: Exception) {
                ""
            }
        }

    val distanceKm: Double
        get() = distance / 1000.0

    companion object {
        /**
         * AFAD JSON API'den parse
         */
        fun fromJsonAfad(json: Map<String, Any?>): Earthquake {
            var lat = 0.0
            var lng = 0.0

            val geojson = json["geojson"] as? Map<*, *>
            val coords = geojson?.get("coordinates") as? List<*>
            if (coords != null && coords.size >= 2) {
                lng = parseDouble(coords[0])
                lat = parseDouble(coords[1])
            } else {
                lat = parseDouble(json["lat"] ?: json["latitude"])
                lng = parseDouble(json["lng"] ?: json["longitude"])
            }

            val dateStr = (json["date_time"] ?: json["date"])?.toString() ?: ""
            val dateMs = parseAfadDateToEpochMs(dateStr)

            return Earthquake(
                id = (json["earthquake_id"] ?: json["id"] ?: "").toString(),
                title = (json["title"] ?: json["location"] ?: "Bilinmeyen Konum").toString(),
                dateEpochMs = dateMs,
                mag = parseDouble(json["mag"] ?: json["magnitude"]),
                depth = parseDouble(json["depth"]),
                latitude = lat,
                longitude = lng,
            )
        }

        /**
         * USGS GeoJSON API'den parse
         */
        fun fromJsonUSGS(json: Map<String, Any?>): Earthquake {
            val properties = json["properties"] as? Map<*, *> ?: emptyMap<String, Any?>()
            val geometry = json["geometry"] as? Map<*, *> ?: emptyMap<String, Any?>()
            val coordinates = geometry["coordinates"] as? List<*> ?: listOf(0.0, 0.0, 0.0)

            val timeMs = (properties["time"] as? Number)?.toLong() ?: System.currentTimeMillis()

            var title = (properties["place"] ?: "Bilinmeyen Konum").toString()
            if (title.contains(" of ")) {
                title = title.substringAfter(" of ").trim()
            }

            return Earthquake(
                id = (json["id"] ?: "").toString(),
                title = title,
                dateEpochMs = timeMs,
                mag = parseDouble(properties["mag"]),
                depth = parseDouble(if (coordinates.size > 2) coordinates[2] else 0.0),
                latitude = parseDouble(coordinates.getOrNull(1) ?: 0.0),
                longitude = parseDouble(coordinates.getOrNull(0) ?: 0.0),
            )
        }

        /**
         * Kandilli RSS XML'den parse.
         * title: "1.5 (ML) MALAZGIRT"
         * description: date, lat, lng, depth bilgilerini içerir
         */
        fun fromKandilliXml(
            titleText: String,
            descText: String,
            linkText: String
        ): Earthquake {
            // Büyüklük
            var mag = 0.0
            val magMatch = Regex("""^([\d.]+)""").find(titleText)
            if (magMatch != null) {
                mag = magMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            }

            // Konum adı
            val location = titleText.replace(Regex("""^[\d.]+\s*(?:\([a-zA-Z]+\)|\-)?\s*"""), "").trim()

            // Tarih (description'dan): "2024.12.11 09:22:06"
            var dateMs = System.currentTimeMillis()
            val dateMatch = Regex("""(\d{4})[./-](\d{2})[./-](\d{2})\s+(\d{2}):(\d{2}):(\d{2})""").find(descText)
            if (dateMatch != null) {
                try {
                    val (y, mo, d, h, mi, s) = dateMatch.destructured
                    // Basit epoch hesabı (yaklaşık)
                    val dateStr = "$y-${mo}-${d}T${h}:${mi}:${s}"
                    val instant = Instant.parse("${dateStr}Z")
                    dateMs = instant.toEpochMilliseconds()
                } catch (_: Exception) {}
            }

            // Derinlik
            var depth = 0.0
            val depthMatch = Regex("""Derinlik[^\d]*([\d.]+)""").find(descText)
            if (depthMatch != null) {
                depth = depthMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            }

            // Koordinatlar (float'ları yakala)
            var lat = 0.0
            var lng = 0.0
            val floatMatches = Regex("""\b(\d+\.\d+)\b""").findAll(descText).toList()
            if (floatMatches.size >= 3) {
                lat = floatMatches[1].groupValues[1].toDoubleOrNull() ?: 0.0
                lng = floatMatches[2].groupValues[1].toDoubleOrNull() ?: 0.0
            }
            if (depth == 0.0 && floatMatches.isNotEmpty()) {
                depth = (if (floatMatches.size >= 4) floatMatches[3].groupValues[1] else floatMatches.last().groupValues[1]).toDoubleOrNull() ?: 0.0
            }

            val id = if (linkText.isNotEmpty()) {
                linkText.split("/").last().replace(".asp", "")
            } else {
                dateMs.toString()
            }

            return Earthquake(
                id = id,
                title = location.ifEmpty { "Bilinmeyen Konum" },
                dateEpochMs = dateMs,
                mag = mag,
                depth = depth,
                latitude = lat,
                longitude = lng,
            )
        }

        // --- Yardımcı fonksiyonlar ---

        private fun parseDouble(value: Any?): Double {
            return when (value) {
                is Number -> value.toDouble()
                is String -> value.toDoubleOrNull() ?: 0.0
                else -> 0.0
            }
        }

        private fun parseAfadDateToEpochMs(dateStr: String): Long {
            if (dateStr.isBlank()) return System.currentTimeMillis()
            return try {
                val clean = dateStr.trim()
                if (clean.contains('.')) {
                    // "21.03.2024 13:00:24" veya "2024.03.21 13:00:24"
                    val parts = clean.split(' ')
                    val dateParts = parts[0].split('.')
                    val timeParts = parts.getOrElse(1) { "00:00:00" }.split(':')
                    val (y, m, d) = if (dateParts[0].length == 4) {
                        Triple(dateParts[0].toInt(), dateParts[1].toInt(), dateParts[2].toInt())
                    } else {
                        Triple(dateParts[2].toInt(), dateParts[1].toInt(), dateParts[0].toInt())
                    }
                    val h = timeParts.getOrElse(0) { "0" }.toInt()
                    val mi = timeParts.getOrElse(1) { "0" }.toInt()
                    val s = timeParts.getOrElse(2) { "0" }.toInt()
                    val isoStr = "$y-${m.toString().padStart(2, '0')}-${d.toString().padStart(2, '0')}T${h.toString().padStart(2, '0')}:${mi.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}Z"
                    Instant.parse(isoStr).toEpochMilliseconds()
                } else if (clean.contains('-') && clean.contains(' ') && !clean.contains('T')) {
                    // "2026-09-02 17:43:34"
                    val isoStr = clean.replace(' ', 'T') + "Z"
                    Instant.parse(isoStr).toEpochMilliseconds()
                } else {
                    Instant.parse(clean).toEpochMilliseconds()
                }
            } catch (_: Exception) {
                System.currentTimeMillis()
            }
        }

        /**
         * İki koordinat arası mesafeyi metre cinsinden hesaplar (Haversine formülü)
         */
        fun haversineDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
            val r = 6371000.0 // Dünya yarıçapı (metre)
            val dLat = (lat2 - lat1) * (PI / 180.0)
            val dLon = (lon2 - lon1) * (PI / 180.0)
            val a = sin(dLat / 2).pow(2) +
                    cos(lat1 * (PI / 180.0)) * cos(lat2 * (PI / 180.0)) *
                    sin(dLon / 2).pow(2)
            val c = 2 * atan2(sqrt(a), sqrt(1 - a))
            return r * c
        }

        /**
         * Depremleri belirli bir yarıçap içinde filtreler
         */
        fun filterWithinRadius(
            earthquakes: List<Earthquake>,
            centerLat: Double,
            centerLng: Double,
            radiusKm: Double = 100.0
        ): List<Earthquake> {
            return earthquakes.filter { eq ->
                val dist = haversineDistance(centerLat, centerLng, eq.latitude, eq.longitude)
                eq.distance = dist
                dist <= radiusKm * 1000
            }
        }

        private object System {
            fun currentTimeMillis(): Long = kotlinx.datetime.Clock.System.now().toEpochMilliseconds()
        }
    }
}
